sklearn/cluster/_kmeans.py

```diff
@@ -245,38 +258,41 @@ def k_means(X, n_clusters, sample_weight=None, init='k-means++',
         an int to make the randomness deterministic.
         See :term:`Glossary <random_state>`.
 
-    copy_x : bool, optional
+    copy_x : bool, default=True
         When pre-computing distances it is more numerically accurate to center
-        the data first.  If copy_x is True (default), then the original data is
-        not modified, ensuring X is C-contiguous.  If False, the original data
-        is modified, and put back before the function returns, but small
-        numerical differences may be introduced by subtracting and then adding
-        the data mean, in this case it will also not ensure that data is
-        C-contiguous which may cause a significant slowdown.
-
-    n_jobs : int or None, optional (default=None)
-        The number of jobs to use for the computation. This works by computing
-        each of the n_init runs in parallel.
-
-        ``None`` means 1 unless in a :obj:`joblib.parallel_backend` context.
-        ``-1`` means using all processors. See :term:`Glossary <n_jobs>`
-        for more details.
-
-    algorithm : "auto", "full" or "elkan", default="auto"
+        the data first. If copy_x is True (default), then the original data is
+        not modified. If False, the original data is modified, and put back
+        before the function returns, but small numerical differences may be
+        introduced by subtracting and then adding the data mean. Note that if
+        the original data is not C-contiguous, a copy will be made even if
+        copy_x is False. If the original data is sparse, but not in CSR format,
+        a copy will be made even if copy_x is False.
+
+    n_jobs : int, default=None
+        The number of OpenMP threads to use for the computation. Parallelism is
+        sample-wise on the main cython loop which assigns each sample to its
+        closest center.
+
+        ``None`` or ``-1`` means using all processors.
```

Comment:
The number of threads used by OpenMP should be controlled via the OMP_NUM_THREADS env variable.

This has been discussed and agreed upon in:

- https://github.com/scikit-learn/scikit-learn/issues/14265
- https://github.com/scikit-learn/scikit-learn/pull/15116
- https://github.com/scikit-learn/scikit-learn/pull/14196
- https://github.com/scikit-learn/scikit-learn/issues/16016
- the core dev meetings

You have been an active part of each of these discussions, so you are perfectly aware of our decision.

I don't understand why you are (again) trying to push for something else here.

Reply:
It's because in this case, the former implementation of K-Means exposed a way to control the amount of parallelism via the `n_jobs` argument. Shall we just remove this feature all at once?

Reply:
@NicolasHug Please be less aggressive in your comments. Before trashing me please ask yourself if I did not do that for a legitimate reason. I don't do stuff just to annoy you. The PR has a lot of comments and I get that reading everything to make sure this point has not been discussed previously is a pain but at least you could ask more nicely. Let me remind you the code of conduct we just explicitly added:
```We are a community based on openness and friendly, didactic, discussions.```

Regarding `n_jobs` the discussion was started here: [comment] (https://github.com/scikit-learn/scikit-learn/pull/11950#discussion_r366936698)

Moreover it was decided in those meetings a general behavior for openmp threads which can be re-evaluated on a case by case. And OMP_NUM_THREADS still controls the number of threads if you let n_jobs=None in this implementation. Anyway, I'm not "pushing" in any direction. I'm just addressing the comments of the reviewers and trying to find a common ground as best as I can.

Reply:
> It's because in this case, the former implementation of K-Means exposed a way to control the amount of parallelism via the n_jobs argument. Shall we just remove this feature all at once?

We can also deprecate it. This is currently changing the default behavior of n_jobs=None, so technically we should have a deprecation of some sort anyway?

@jeremiedbb , I wasn't trashing you, I was being factual. We've discussed this many times before so I left out the sugar coating.

For the sake of openness and discussions, if this PR is going to go against the decided default, then there should be a discussion. An open one. I'm sorry but https://github.com/scikit-learn/scikit-learn/pull/11950#discussion_r366936698 does not qualify.

This PR chooses option (2) from https://github.com/scikit-learn/scikit-learn/issues/14265 , where various core-devs have expressed concerns about it. Option (2) was *precisely* the reason https://github.com/scikit-learn/scikit-learn/issues/14265 was opened. It was decided not to use it in general.

Using it back without further consultation is dismissing the long discussions and effots we've all put into this.

Reply:
From what I remember, the outcome of the team meeting was, by default, to not expose a control for the number of OpenMP threads whenever we use Cython prange but stay open for specific discussions on a case by case basis in specific PRs.

But alright let's deprecate the n_jobs parameter in KMeans and just document in the docstring than OpenMP-based parallelism and maybe link to https://scikit-learn.org/dev/modules/computing.html#openmp-based-parallelism in the main doc.

Reply:
Yes. I'm of course still open to discuss this. I was frustrated not to see such discussion happening  (I did look for one, here and in the other places). Though I should have just said that instead of writing a harsh comment.

FWIW, I still think that for KMeans and HistGradientBoosting, the best alternative is to have a `n_omp_threads` parameter + top priority to OMP_NUM_THREADS. The main criticism I read against it is that it is exposing an implementation detail, but so does (2) in a more insidious way: different default behavior + incompatibility with joblib context managers.
So I still haven't been convinced that this isn't the right choice.

Reply:
If KMeans didn't have a `n_jobs` param we would not add a new parameter, according to #14265. So I'm not in favor of deprecating `n_jobs` to add `n_omp_threads`. I agree that having a `n_jobs` parameter with a different meaning and default is confusing, so I'm more in favor of deprecating `n_jobs`.

Reply:
I am in favor of deprecating n_jobs for now. And maybe later adding a dedicated `n_omp_threads` param (or similar) if experience shows that setting `OMP_NUM_THREADS` env var is too cumbersome / limiting.